extensions/meme_rater.py

```diff
@@ -0,0 +1,64 @@
+import os
+import hikari, lightbulb
+from PIL import Image
+import requests
+from snark import model
+from google.generativeai.generative_models import GenerativeModel
+from typing import Final
+
+plugin = lightbulb.Plugin("MemeRater")
+
+MEMES_CHANNEL_ID = os.environ.get("MEME_CHANNEL_ID")
+
+# Make this envs if you want
+MEME_RATE_PROMPT: Final[str] = (
+    "Rate this meme out of 10, with 10 being the funniest. Rate is solely on how funny a bunch of computer science nerds would find it. ONLY Return an integer."
+)
+MINIMUM_MEME_RATING_TO_NOT_DELETE: Final[int] = 6  # can be in .env if you prefer
+IMG_FILE_EXTENSIONS: Final = {"jpg", "jpeg", "png", "webp"}
+# removed uncommon formats for testing - don't tell Jimmy plx
+# "tiff",  "bmp", "gif <-- idk if these are supported
+
+
+async def get_meme_rating(image_url: str, model: GenerativeModel):
+    response = model.generate_content(
+        contents=[MEME_RATE_PROMPT, Image.open(requests.get(image_url).raw)]
+    )
+    return response.text
+
+
+@plugin.listener(hikari.GuildMessageCreateEvent)
+async def main(event: hikari.GuildMessageCreateEvent) -> None:
+    if event.channel_id != MEMES_CHANNEL_ID:
+        return
+    for attachment in event.message.attachments:
+        att_ext = attachment.extension
+        if att_ext in IMG_FILE_EXTENSIONS:
+            image_url = attachment.url
+            res = await get_meme_rating(image_url, model)
+            if res:
```

Comment:
Please never nest this

Reply:
This will pause waiting for the result sequentially if there are multiple attachments? I assume there's a way to fire off multiple rating requests in parallel.

Note: I am not sure how multiple attachments actually happen (can they be forwarded all in one message?) in Discord.

Reply:
You raise a great point. I think we should consider multithreading before merging.

Reply:
I can't remember if it's pleasant to do in Python, though. Would avoid if it's not easy.

Reply:
`asyncio.gather` is the python equivalent of js `Promise.all`.

However:
- the genai model has a rate limit of 1 request per 2 seconds
- the function `get_meme_rating` is completely synchronous so it's useless to await it